File: frontend-react/src/lib/api/convert.ts
import {
  postFile,
  postMultipart,
} from "./client";

/**
 * Konversi JPG ke PDF.
 */
export function jpgToPdf(
  files: File | Blob | Array<File | Blob>,
): Promise<Blob> {
  const values = Array.isArray(files) ? files : [files];
  const formData = new FormData();

  for (const file of values) {
    formData.append("file", file);
  }

  return postMultipart("/api/v1/convert/jpg-to-pdf", formData);
}

/**
 * Konversi PDF ke JPG.
 */
export function pdfToJpg(
  file: File,
): Promise<Blob> {
  return postFile(
    "/api/v1/convert/pdf-to-jpg",
    file,
  );
}

/**
 * Konversi Word ke PDF.
 */
export function wordToPdf(
  file: File,
): Promise<Blob> {
  return postFile(
    "/api/v1/convert/word-to-pdf",
    file,
  );
}

/**
 * Konversi Excel ke PDF.
 */
export function excelToPdf(
  file: File,
): Promise<Blob> {
  return postFile(
    "/api/v1/convert/excel-to-pdf",
    file,
  );
}

/**
 * Konversi PowerPoint ke PDF.
 */
export function powerpointToPdf(
  file: File,
): Promise<Blob> {
  return postFile(
    "/api/v1/convert/powerpoint-to-pdf",
    file,
  );
}

/**
 * Konversi PDF ke Word.
 */
export function pdfToWord(
  file: File,
): Promise<Blob> {
  return postFile(
    "/api/v1/convert/pdf-to-word",
    file,
  );
}

/**
 * Konversi PDF ke Excel.
 */
export function pdfToExcel(
  file: File,
): Promise<Blob> {
  return postFile(
    "/api/v1/convert/pdf-to-excel",
    file,
  );
}

File: backend/src/routes/v1/convert.rs
use axum::{Router, routing::post};

use crate::state::AppState;

use crate::features::convert::handlers::{
    excel_to_pdf, jpg_to_pdf, pdf_to_excel, pdf_to_jpg, pdf_to_word, word_to_pdf,
};

/// Endpoint konversi dokumen di API v1.
pub fn routes() -> Router<AppState> {
    Router::<AppState>::new()
        .route("/convert/excel-to-pdf", post(excel_to_pdf::handler))
        .route("/convert/jpg-to-pdf", post(jpg_to_pdf::handler))
        .route("/convert/pdf-to-excel", post(pdf_to_excel::handler))
        .route("/convert/pdf-to-jpg", post(pdf_to_jpg::handler))
        .route("/convert/pdf-to-word", post(pdf_to_word::handler))
        .route("/convert/word-to-pdf", post(word_to_pdf::handler))
}
